### PR TITLE
Retrieve parent window ID for XDG desktop portal dialogs

### LINKED ISSUE #430: Make XDG dialogs modal
**Describe the problem or limitation you are having**

We currently do not set `parent_window` in XDG dialogs:

https://github.com/crow-translate/crow-translate/blob/0d52fa0769ce84cdb2a0b0e7ee21f9f78ae3dd05/src/settings/autostartmanager/portalautostartmanager.cpp#L50-L52
https://github.com/crow-translate/crow-translate/blob/0d52fa0769ce84cdb2a0b0e7ee21f9f78ae3dd05/src/ocr/screengrabbers/waylandportalscreengrabber.cpp#L52-L54

**Describe the solution you'd like**
We should detect parent window and pass it over D-Bus.

### PR DESCRIPTION
**Pull request by gasinvein**
Follow-up from https://github.com/crow-translate/crow-translate/pull/429#issuecomment-1126707432, attempt to fix #430

It doesn't crash anymore even if we don't check if the active window isn't `nullptr`, but I'm still not sure if this way to determine `parent_window` ID for XDP is correct. Looks like it is for X11, but not for Wayland.

### CODE REVIEW

**Comment by Shatur** [comment]
Looks awesome. But are you sure that on your system `QX11Info::isPlatformX11()` returns `true`?

**Comment by Shatur** [comment]
Just asking because in TG they format it as hex on X11: https://github.com/desktop-app/lib_base/blob/5ab3cdaef587b8ac8fdf1a96879961fd57ac7ace/base/platform/linux/base_linux_xdp_utilities.cpp#L32

**Comment by Shatur** [comment]
And we don't do it for x11. So maybe on your machine you having wayland code path?

**Comment by gasinvein** [comment]
> Looks awesome. But are you sure that on your system `QX11Info::isPlatformX11()` returns `true`?

I'm sniffing dbus communications with Bustle. There I can see that Crow Translate sends a `parent_windows` arg starting with `x11:`.

> Just asking because in TG they format it as hex on X11: https://github.com/desktop-app/lib_base/blob/5ab3cdaef587b8ac8fdf1a96879961fd57ac7ace/base/platform/linux/base_linux_xdp_utilities.cpp#L32

Ah yes, now I recall there was some dec/hex issue with Qt and XDP.

> And we don't do it for x11. So maybe on your machine you having wayland code path?

I guess an invalid `parent_window` value is just interpreted as empty. This explains why the dialog still doesn't seem modal for me.

**Comment by Shatur** [comment]
> I guess an invalid parent_window value is just interpreted as empty.

It looks so. Could you check if it works formatted as hex? Just pass two additional parameters to [arg](https://doc.qt.io/qt-5/qstring.html#arg-1): `0` and 16.

**Comment by gasinvein** [comment]
Maybe I'm missing something, but the dialog still doesn't seem to be modal. This seems the case for Telegram as well. Maybe they're doing something incorrect?

**Comment by Shatur** [comment]
I don't know :(
The spec is poor, to be honest.

**Comment by gasinvein** [comment]
Tried adding a `0x` prefix to XID but doesn't seem to change anything. The hex digits sent over dbus seem to be correct, so either we're missing something, or the `parent_window` value doesn't really affect how GNOME presents the dialog.

I'd say leave it as it is here for now, maybe someone will come up with a more valid way way to determine XDP window id.

**Comment by Shatur** [comment]
Okay! Thank you a lot for looking into it. I will draft a new release as is then.

**Comment by gasinvein** [comment]
Left a `FIXME`, so it's visible there is some work to do.

**Comment by gasinvein** [comment]
I was told by a more knowledgeable folk that the Background portal prompt in GNOME is expected to be fullscreen and _not_ modal. So I believe we here got it right for X11 window ID.

**Comment by Shatur** [comment]
> I was told by a more knowledgeable folk that the Background portal prompt in GNOME is expected to be fullscreen and not modal. So I believe we here got it right for the X11 window ID.

Then why `parent_window` is for? On other DE's it's modal?

**Comment by gasinvein** [comment]
> > I was told by a more knowledgeable folk that the Background portal prompt in GNOME is expected to be fullscreen and not modal. So I believe we here got it right for the X11 window ID.
> 
> Then what is `parent_window` for? Is it modal on other DEs?

Some other portal dialogs are modal. And I guess dialogs that aren't modal in GNOME may be modal in other DEs, yes.

**Comment by Shatur** [comment]
Then let's merge it as is?

**Comment by Shatur** [comment]
One question. Will it work if we pass `winId()` from the parent widget directly?

File: src/xdgdesktopportal.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#define XDGDESKTOPPORTAL_H
+
+#include <QString>
+#include <QWindow>
+
+namespace XdgDesktopPortal
+{
+// Retrieve parent window in string form according to
+// https://flatpak.github.io/xdg-desktop-portal/#parent_window
+QString parentWindow(QWindow *activeWindow);
```
Comment (Shatur) [inline src/xdgdesktopportal.h:32]: I.e accept `winID` here.

File: src/settings/autostartmanager/portalautostartmanager.cpp [hunk 2]
```diff
@@ -41,15 +43,14 @@ bool PortalAutostartManager::isAutostartEnabled() const
 
 void PortalAutostartManager::setAutostartEnabled(bool enabled)
 {
+    auto *window = qobject_cast<QWidget *>(parent())->windowHandle();
```
Comment (Shatur) [inline src/settings/autostartmanager/portalautostartmanager.cpp:46]: And do it like this
```suggestion
    auto *winId = qobject_cast<QWidget *>(parent())->winId();
```

**Review by Shatur — CHANGES_REQUESTED** [review]

**Comment by gasinvein** [comment]
> Then let's merge it as is?

Yeah, I believe it won't introduce regressions at least. Probably I (or someone else) will add a proper wayland window ID implementation on top of this PR.

> One question. Will it work if we pass `winId()` from the parent widget directly?

I think not, we should pass a `QWindow` objects into to function, since the algorithm for Wayland is entirely different (it won't use `winId()` at all).

**Comment by Shatur** [comment]
Oh, I see: https://api.kde.org/frameworks/kwayland/html/classKWayland_1_1Client_1_1Surface.html#a0b2321cb22b1dbce2903cdac1346a3d2
But there is such method from winId: https://api.kde.org/frameworks/kwayland/html/classKWayland_1_1Client_1_1Surface.html#a471ecec2d7698a90edbebe27ad471933

File: src/xdgdesktopportal.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+#include "xdgdesktopportal.h"
+
+#include <QX11Info>
+
+QString XdgDesktopPortal::parentWindow(QWindow *activeWindow)
+{
+    const QString platformName = QX11Info::isPlatformX11() ? QStringLiteral("x11") : QStringLiteral("wayland");
+    // FIXME this is most likely incorrect, but we can't figure out the correct format
```
Comment (Shatur) [inline src/xdgdesktopportal.cpp:29]: Since on wayland it should be a surface let's print a `qWarning` an return empty string.

**Comment by gasinvein** [comment]
> Oh, I see: https://api.kde.org/frameworks/kwayland/html/classKWayland_1_1Client_1_1Surface.html#a0b2321cb22b1dbce2903cdac1346a3d2
> But there is such method from winId: https://api.kde.org/frameworks/kwayland/html/classKWayland_1_1Client_1_1Surface.html#a471ecec2d7698a90edbebe27ad471933

We can't use these methods unless we add `libkwin` dependency, I suppose.

**Comment by Shatur** [comment]
> We can't use these methods unless we add libkwin dependency, I suppose.

Agree, that's just what Telegram use.
So we can use WinId directly.

File: src/xdgdesktopportal.cpp (hunk 3 above)
Comment (gasinvein) [reply src/xdgdesktopportal.cpp]: Done.

**Comment by Shatur** [comment]
Simplified things a little. Good to merge?

**Review by Shatur — APPROVED** [review]

**Comment by gasinvein** [comment]
> Simplified things a little. Good to merge?

Should be. Although I still believe we should pass the `QWindow`. The KWayland implementation of `Surface::fromQtWinId` actually does backward conversion from `WId` to `QWindow`:
```c++
 Surface *Surface::fromQtWinId(WId wid)
 {
     QWindow *window = nullptr;
  
     for (auto win : qApp->allWindows()) {
         if (win->winId() == wid) {
             window = win;
             break;
         }
     }
  
     if (!window) {
         return nullptr;
     }
     return fromWindow(window);
 }
```
So we need a `QWindow` anyway.

**Comment by Shatur** [comment]
Oh, good catch. Let me fix it.

**Comment by gasinvein** [comment]
> Good to merge?

Looks good.

**Comment by Shatur** [comment]
Thank you a lot!

**Comment by gasinvein** [comment]
A side note for the future: in `SettingsDialog::accept`, we should probably call `QDialog::accept()` _after_ invoking the autostart manager; otherwise the window would be already closed when it reaches `XdgDesktopPortal::parentWindow`, so we won't be able to access its wayland surface.

**Comment by Shatur** [comment]
It should be normal for windows not to close immediately, as far as I know. They close at the end of the event loop.

**Comment by gasinvein** [comment]
> windows don't close immediately, as far as I know. They close at the end of the event loop.

Maybe it's machine-dependent? I've got impression that currently we reach autostart manager when the window is closed.

**Comment by Shatur** [comment]
Oh, I think it's because we allocating a new event loop inside autostart manager to wait for response.
So in your case the settings window is closed before the XDG dialog?

**Comment by gasinvein** [comment]
> So in your case the settings window is closed before the XDG dialog?

I'm not completely sure but it looks very much like it. If I put a, say, 5 seconds delay at the end of `PortalAutostartManager::setAutostartEnabled`, it's pretty visible that the settings window is closed immediately after the "OK" button is clicked, while the main window stays unresponsive for that 5 seconds.

**Comment by Shatur** [comment]
Interesting. Does moving `QDialog::accept` helps?

**Comment by gasinvein** [comment]
> Does moving `QDialog::accept` down helps?

It does. If I move it to the end of `SettingsDialog::accept`, the delay happens while the settings window is still open.

**Comment by Shatur** [comment]
Could you send this change too? :)

**Comment by gasinvein** [comment]
Sure, see the link above.

**Pull request merged**